File: frontend/mobile/lib/service/kal_service.dart
import 'dart:convert';
import '../config/api_config.dart';
import 'storage_helper.dart';
import 'auth_service.dart';
import '../models/kal_models.dart';

// Service สำหรับจัดการข้อมูลแคลอรี่
class KalService {
  static String get baseUrl => ApiConfig.dailyUrl;

  // คำนวณและบันทึกแคลอรี่ (BMR + TDEE + Target)
  static Future<CalculateCaloriesResponse> calculateAndSaveCalories({
    required double activityLevel,
  }) async {
    try {
      final userId = await StorageHelper.getUserId();
      if (userId == null) {
        throw Exception('User ID not found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/calculate-calories/$userId');

      final response = await AuthService.authenticatedRequest(
        method: 'POST',
        endpoint: url.toString(),
        body: jsonEncode({'activityLevel': activityLevel}),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return CalculateCaloriesResponse.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to calculate calories');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('Error calculating calories: $e');
    }
  }

  // ดึงสถานะแคลอรี่ (Target, Consumed, Burned, Net, Remaining)
  static Future<CalorieStatus> getCalorieStatus() async {
    try {
      final userId = await StorageHelper.getUserId();
      if (userId == null) {
        throw Exception('User ID not found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/status/$userId');

      final response = await AuthService.authenticatedRequest(
        method: 'GET',
        endpoint: url.toString(),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return CalorieStatus.fromJson(data);
      } else if (response.statusCode == 404) {
        return CalorieStatus(
          activityLevel: 0,
          targetCalories: 0,
          consumedCalories: 0,
          burnedCalories: 0,
          netCalories: 0,
          remainingCalories: 0,
        );
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch calorie status');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('Error fetching calorie status: $e');
    }
  }

  // ดึงข้อมูลสารอาหาร (Protein, Fat, Carbohydrate)
  static Future<DailyMacros> getDailyMacros() async {
    try {
      final userId = await StorageHelper.getUserId();
      if (userId == null) {
        throw Exception('User ID not found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/macros/$userId');

      final response = await AuthService.authenticatedRequest(
        method: 'GET',
        endpoint: url.toString(),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return DailyMacros.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch daily macros');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('Error fetching daily macros: $e');
    }
  }

  // ดึงข้อมูลแคลอรี่รายสัปดาห์
  static Future<WeeklyCaloriesResponse> getWeeklyCalories() async {
    try {
      final userId = await StorageHelper.getUserId();
      if (userId == null) {
        throw Exception('User ID not found. Please login again.');
      }

      final url = Uri.parse('$baseUrl/weekly/$userId');

      final response = await AuthService.authenticatedRequest(
        method: 'GET',
        endpoint: url.toString(),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return WeeklyCaloriesResponse.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch weekly calories');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('Error fetching weekly calories: $e');
    }
  }
}

File: frontend/web/src/services/kal_service.ts
import api from './auth_service'
import { getNodeApiUrl } from '@/config/api.config'

const API_BASE_URL = getNodeApiUrl()
const DAILY_API_URL = `${API_BASE_URL}/api/daily`

// ประเภทข้อมูล Response และ Model
export interface CalculateCaloriesResponse {
  message: string
  activity_level: number
  bmr: number
  tdee: number
  target_calories: number
  goal: string
}

export interface CalorieStatus {
  activity_level: number
  target_calories: number
  consumed_calories: number
  burned_calories: number
  net_calories: number
  remaining_calories: number
}

export interface DailyMacros {
  message: string
  protein: number
  fat: number
  carbohydrate: number
}

export interface DailyCalorieData {
  date: string
  net_calories: number
}

export interface WeeklyCaloriesResponse {
  message: string
  data: DailyCalorieData[]
}

// ดึง User ID จาก localStorage
function getUserId(): string | null {
  if (typeof window === 'undefined') {
    return null;
  }

  const userStr = localStorage.getItem('user')
  if (!userStr) {
    return null;
  }

  try {
    const user = JSON.parse(userStr)
    if (!user.id) {
      return null;
    }
    return user.id?.toString() || null
  } catch (error) {
    return null
  }
}

// Service สำหรับคำนวณและจัดการแคลอรี่
export const kalService = {
  // คำนวณและบันทึก BMR, TDEE และแคลอรี่เป้าหมาย
  calculateAndSaveCalories: async (activityLevel: number): Promise<CalculateCaloriesResponse> => {
    try {
      const userId = getUserId()
      if (!userId) {
        throw new Error('User ID not found. Please login again.')
      }

      const url = `${DAILY_API_URL}/calculate-calories/${userId}`
      const response = await api.post<CalculateCaloriesResponse>(url, {
        activityLevel: activityLevel,
      })

      return response.data
    } catch (error: any) {
      const errorMessage = error.response?.data?.message || error.message || 'Failed to calculate calories'
      throw new Error(errorMessage)
    }
  },

  // ดึงสถานะแคลอรี่ (เป้าหมาย, บริโภค, เผาผลาญ, คงเหลือ)
  getCalorieStatus: async (): Promise<CalorieStatus> => {
    try {
      const userId = getUserId()
      if (!userId) {
        throw new Error('User ID not found. Please login again.')
      }

      const url = `${DAILY_API_URL}/status/${userId}`
      const response = await api.get<CalorieStatus>(url)

      return response.data
    } catch (error: any) {
      if (error.response?.status === 404) {
        return {
          activity_level: 0,
          target_calories: 0,
          consumed_calories: 0,
          burned_calories: 0,
          net_calories: 0,
          remaining_calories: 0,
        }
      }

      const errorMessage = error.response?.data?.message || error.message || 'Failed to fetch calorie status'
      throw new Error(errorMessage)
    }
  },

  // ดึงข้อมูลสารอาหารรายวัน (โปรตีน, ไขมัน, คาร์โบไฮเดรต)
  getDailyMacros: async (): Promise<DailyMacros> => {
    try {
      const userId = getUserId()
      if (!userId) {
        throw new Error('User ID not found. Please login again.')
      }

      const url = `${DAILY_API_URL}/macros/${userId}`
      const response = await api.get<DailyMacros>(url)

      return response.data
    } catch (error: any) {
      const errorMessage = error.response?.data?.message || error.message || 'Failed to fetch daily macros'
      throw new Error(errorMessage)
    }
  },

  // ดึงข้อมูลแคลอรี่รายสัปดาห์ (7 วันล่าสุด)
  getWeeklyCalories: async (): Promise<WeeklyCaloriesResponse> => {
    try {
      const userId = getUserId()
      if (!userId) {
        throw new Error('User ID not found. Please login again.')
      }

      const url = `${DAILY_API_URL}/weekly/${userId}`
      const response = await api.get<WeeklyCaloriesResponse>(url)

      return response.data
    } catch (error: any) {
      if (error.response?.status === 401) {
        throw new Error('Session expired. Please login again.')
      } else if (error.response?.status === 404) {
        return {
          message: 'No weekly data found',
          data: []
        }
      }

      const errorMessage = error.response?.data?.message || error.message || 'Failed to fetch weekly calories'
      throw new Error(errorMessage)
    }
  },
}

export default kalService
